includes/reader-activation/class-esp-sync.php

```diff
@@ -0,0 +1,202 @@
+	 *
+	 * @return true|\WP_Error True if succeeded or WP_Error.
+	 */
+	protected static function sync( $contact, $context = '' ) {
+		$can_sync = static::can_esp_sync( true );
+		if ( $can_sync->has_errors() ) {
+			return $can_sync;
+		}
+
+		if ( empty( $context ) ) {
+			$context = static::$context;
+		}
+
+		$master_list_id = Reader_Activation::get_esp_master_list_id();
+
+		/**
+		 * Filters the contact data before syncing to the ESP.
+		 *
+		 * @param array  $contact The contact data to sync.
+		 * @param string $context The context of the sync.
+		 */
+		$contact = \apply_filters( 'newspack_esp_sync_contact', $contact, $context );
+
+		$result = \Newspack_Newsletters_Contacts::upsert( $contact, $master_list_id, $context );
+
+		return \is_wp_error( $result ) ? $result : true;
+	}
+
+	/**
+	 * Given a user ID or WooCommerce Order, sync that reader's contact data to
+	 * the connected ESP.
+	 *
+	 * @param int|\WC_order $user_id_or_order User ID or WC_Order object.
+	 * @param bool          $is_dry_run       True if a dry run.
+	 *
+	 * @return true|\WP_Error True if the contact was synced successfully, WP_Error otherwise.
+	 */
+	protected static function sync_contact( $user_id_or_order, $is_dry_run = false ) {
+		$can_sync = static::can_esp_sync( true );
+		if ( ! $is_dry_run && $can_sync->has_errors() ) {
+			return $can_sync;
+		}
+
+		$is_order = $user_id_or_order instanceof \WC_Order;
+		$order    = $is_order ? $user_id_or_order : false;
+		$user_id  = $is_order ? $order->get_customer_id() : $user_id_or_order;
+		$user     = \get_userdata( $user_id );
+
+		// Backfill Network Registration Site field if needed.
+		$registration_site = false;
+		if ( $user && defined( 'NEWSPACK_NETWORK_READER_ROLE' ) && defined( 'Newspack_Network\Utils\Users::USER_META_REMOTE_SITE' ) ) {
+			if ( ! empty( array_intersect( $user->roles, \Newspack_Network\Utils\Users::get_synced_user_roles() ) ) ) {
+				$registration_site = \esc_url( \get_site_url() ); // Default to current site.
+				$remote_site       = \get_user_meta( $user->ID, \Newspack_Network\Utils\Users::USER_META_REMOTE_SITE, true );
+				if ( ! empty( \wp_http_validate_url( $remote_site ) ) ) {
+					$registration_site = \esc_url( $remote_site );
+				}
+			}
+		}
+
+		$customer = new \WC_Customer( $user_id );
+		if ( ! $customer || ! $customer->get_id() ) {
+			return new \WP_Error(
+				'newspack_esp_sync_contact',
+				sprintf(
+				// Translators: %d is the user ID.
+					__( 'Customer with ID %d does not exist.', 'newspack-plugin' ),
+					$user_id
+				)
+			);
+		}
+
+		// Ensure the customer has a billing address.
+		if ( ! $customer->get_billing_email() && $customer->get_email() ) {
+			$customer->set_billing_email( $customer->get_email() );
+			$customer->save();
+		}
+
+		$contact = $is_order ? WooCommerce_Connection::get_contact_from_order( $order ) : WooCommerce_Connection::get_contact_from_customer( $customer );
```

Comment:
Seems that all the decisions of what metadata to sync is done in `get_contact_from_order`. This is one of the things you are getting out of there and bringing to the scope of sync, right?

Reply:
Yep, you can see [here](https://github.com/Automattic/newspack-plugin/pull/3362/files#diff-c1aa6e451da797c3ebd6c6fa112527eefc89443140e173e276d25af8adc135b8) that the draft PR proposes a WooCommerce class inside the Sync namespace for handling contact generation based on Woo data.